modules/renter/download.go

```diff
@@ -1,805 +1,505 @@
-
-	// mark the chunks as not being downloaded yet
-	for i := minChunk; i <= maxChunk; i++ {
-		d.finishedChunks[i] = false
-	}
-
-	d.initPieceSet(f, r)
-	return d
-}
-
-// newDownload creates a newly initialized download.
-func newDownload(f *file, destination modules.DownloadWriter) *download {
-	return &download{
-		startTime:        time.Now(),
-		chunkSize:        f.chunkSize(),
-		destination:      destination,
-		erasureCode:      f.erasureCode,
-		fileSize:         f.size,
-		masterKey:        f.masterKey,
-		numChunks:        f.numChunks(),
-		siapath:          f.name,
-		downloadFinished: make(chan struct{}),
-		finishedChunks:   make(map[uint64]bool),
-	}
-}
-
-// initPieceSet initializes the piece set, including calculations of the total download size.
-func (d *download) initPieceSet(f *file, r *Renter) {
-	// Allocate the piece size and progress bar so that the download will
-	// finish at exactly 100%. Due to rounding error and padding, there is not
-	// a strict mapping between 'progress' and 'bytes downloaded' - it is
-	// actually necessary to download more bytes than the size of the file.
-	// The effective size of the download is determined by the number of chunks
-	// to be downloaded. TODO: Handle variable-size last chunk - Same in downloadqueue.go
-	numChunks := uint64(len(d.finishedChunks))
-
-	dlSize := d.length
-	d.reportedPieceSize = dlSize / (numChunks * uint64(d.erasureCode.MinPieces()))
-	d.atomicDataReceived = dlSize - (d.reportedPieceSize * numChunks * uint64(d.erasureCode.MinPieces()))
-
-	// Assemble the piece set for the download.
-	d.pieceSet = make(map[uint64]map[types.FileContractID]pieceData)
-	for i := range d.finishedChunks {
-		d.pieceSet[i] = make(map[types.FileContractID]pieceData)
-	}
-
-	f.mu.RLock()
-	for _, contract := range f.contracts {
-		id := r.hostContractor.ResolveID(contract.ID)
-		for i := range contract.Pieces {
-			// Only add pieceSet entries for chunks that are going to be downloaded.
-			m, exists := d.pieceSet[contract.Pieces[i].Chunk]
-			if exists {
-				m[id] = contract.Pieces[i]
-			}
-		}
-	}
-	f.mu.RUnlock()
-}
-
-// Err returns the error encountered by a download, if it exists.
-func (d *download) Err() error {
+// managedFail will mark the download as complete, but with the provided error.
+// If the download has already failed, the error will be updated to be a
+// concatenation of the previous error and the new error.
+func (d *download) managedFail(err error) {
 	d.mu.Lock()
 	defer d.mu.Unlock()
 
-	return d.downloadErr
-}
-
-// fail will mark the download as complete, but with the provided error.
-func (d *download) fail(err error) {
-	if d.downloadComplete {
-		// Either the download has already succeeded or failed, nothing to do.
+	// If the download is already complete, extend the error.
+	complete := d.staticComplete()
+	if complete && d.err != nil {
+		d.err = errors.Compose(d.err, err)
+		return
```

Comment:
I was thinking that it's better to have more information about which chunks failed. I guess it's not that useful if you know 14 chunks failed instead of 1.